integration_tests/test_runner/src/tests/dex.rs

```diff
@@ -0,0 +1,604 @@
+};
+use althea_proto::cosmos_sdk_proto::cosmos::params::v1beta1::{
+    ParamChange, ParameterChangeProposal,
+};
+use clarity::{Address as EthAddress, PrivateKey, Uint256};
+use deep_space::{Coin, Contact};
+use num::ToPrimitive;
+
+use web30::client::Web3;
+use web30::jsonrpc::error::Web3Error;
+use web30::types::TransactionResponse;
+
+lazy_static! {
+    pub static ref POOL_IDX: Uint256 = 36000u32.into();
+}
+
+pub async fn dex_test(
+    contact: &Contact,
+    web3: &Web3,
+    validator_keys: Vec<ValidatorKeys>,
+    evm_user_keys: Vec<EthermintUserKey>,
+    erc20_contracts: Vec<EthAddress>,
+    dex_contracts: DexAddresses,
+) {
+    let evm_user = evm_user_keys.first().unwrap();
+    let evm_privkey = evm_user.eth_privkey;
+    let optional_caller = Some(evm_user.eth_address);
+    let croc_query_contract = dex_contracts.query;
+    let dex_result = croc_query_dex(web3, croc_query_contract, optional_caller).await;
+    assert!(dex_result.is_ok(), "Bad result");
+    let dex = dex_result.unwrap();
+    assert_eq!(dex, dex_contracts.dex, "Dex contract address mismatch");
+
+    let tokens = erc20_contracts[0..2].to_vec();
+    let (pool_base, pool_quote) = if tokens[0] < tokens[1] {
+        (tokens[0], tokens[1])
+    } else {
+        (tokens[1], tokens[0])
+    };
+
+    let safe_mode = dex_query_safe_mode(web3, dex, Some(evm_user.eth_address))
+        .await
+        .expect("Unable to query safe mode");
+    if safe_mode {
+        info!("Dex is in safe mode, disabling safe mode");
+        submit_and_pass_safe_mode_proposal(contact, &validator_keys, false, true).await;
+    }
+
+    let pool = croc_query_pool_params(
+        web3,
+        dex_contracts.query,
+        Some(evm_user.eth_address),
+        pool_base,
+        pool_quote,
+        *POOL_IDX,
+    )
+    .await;
+    if pool.is_ok() && pool.unwrap().tick_size != 0 {
+        info!("Pool already created, approving use of base and quote tokens");
+        web3.approve_erc20_transfers(pool_base, evm_privkey, dex, Some(OPERATION_TIMEOUT), vec![])
+            .await
+            .expect("Could not approve base token");
+        web3.approve_erc20_transfers(
+            pool_quote,
+            evm_privkey,
+            dex,
+            Some(OPERATION_TIMEOUT),
+            vec![],
+        )
+        .await
+        .expect("Could not approve base token");
+    } else {
+        info!("Creating pool");
+        init_pool(
+            web3,
+            evm_user.eth_privkey,
+            dex_contracts.dex,
+            pool_base,
+            pool_quote,
+            None,
+            None,
+        )
+        .await
+        .expect("Could not create pool");
+    }
+
+    let tick = croc_query_curve_tick(
+        web3,
+        dex_contracts.query,
+        Some(evm_user.eth_address),
+        pool_base,
+        pool_quote,
+        *POOL_IDX,
+    )
+    .await
+    .expect("Could not get curve tick for pool");
+
+    let price = croc_query_price(
```

Comment:
do we try submitting an invalid proposal to ensure it fails? for example a proposal from the wrong address?